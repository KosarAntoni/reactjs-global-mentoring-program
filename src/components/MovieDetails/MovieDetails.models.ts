export interface MovieDetailsProps {
  title: string
  vote_average: number
  release_date: string
  poster_path: string
  overview: string
  genres: string[]
  runtime: number
  className?: string
}
