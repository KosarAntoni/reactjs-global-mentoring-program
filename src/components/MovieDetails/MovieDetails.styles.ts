import styled from '@emotion/styled'

export const StyledMovieDetails = styled.div`
  display: flex;
  flex-direction: row;

  color: var(--primary-text);

  & .heading {
    display: flex;
    flex-direction: row;
    align-items: center;

    h2 {
      margin: 0;

      font-size: 4rem;
      font-weight: 300;
      text-transform: uppercase;
    }

    span {
      width: 6rem;
      height: 6rem;
      margin-left: 2.6rem;

      font-size: 2rem;
      line-height: 6rem;
      text-align: center;
      vertical-align: middle;

      border: 1px solid var(--primary-text);
      border-radius: 10rem;
    }
  }

  & .genres {
    margin-top: 0;

    font-size: 1.4rem;

    opacity: 0.5;
  }

  & .information {
    display: inline-grid;
    grid-template-columns: auto auto;
    column-gap: 5rem;

    font-size: 2.4rem;
    color: var(--pastele-red);
  }

  & .overview {
    font-size: 2rem;

    opacity: 0.5;
  }

  img {
    width: 32rem;
    margin-right: 5.8rem;
  }
`
