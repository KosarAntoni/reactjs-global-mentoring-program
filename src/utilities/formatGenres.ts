export const formatGenres = (genres: string[]): string => genres.length > 2 ? genres.join(', ') : genres.join(' & ')
