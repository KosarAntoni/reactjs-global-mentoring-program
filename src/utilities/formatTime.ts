export const formatTime = (minutes: number): string => {
  const hours = Math.floor(minutes / 60)
  const minutesLeft = minutes % 60
  return `${hours < 10 ? '0' : ''}${hours}h ${
      minutesLeft < 10 ? '0' : ''
    }${minutesLeft} min`
}
