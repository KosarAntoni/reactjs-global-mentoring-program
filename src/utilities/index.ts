export * from './formatGenres'
export * from './formatTime'
