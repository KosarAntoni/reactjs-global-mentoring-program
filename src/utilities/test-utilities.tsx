import React, { PropsWithChildren } from 'react'
import { Provider } from 'react-redux'
import type { PreloadedState } from '@reduxjs/toolkit'
import { configureStore } from '@reduxjs/toolkit'
import type { RenderOptions } from '@testing-library/react'
import { render } from '@testing-library/react'
import type { AppStore, RootState } from 'store'
import moviesSlice, { initialState } from 'store/moviesSlice'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProviders (
  ui: React.ReactElement,
  {
    preloadedState = {
        movies: initialState
    },
    store = configureStore({ reducer: { movies: moviesSlice }, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper ({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    )
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
