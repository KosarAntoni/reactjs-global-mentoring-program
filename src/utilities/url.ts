export const appendToParams = (searchParams: URLSearchParams, newValue?: { [n: string]: string | null }) => {
    const params: any = {}
    searchParams.forEach((value: string | null, key: string) => {
        params[key] = value
    })

    return ({ ...params, ...newValue })
}

export const removeFromParams = (searchParams: URLSearchParams, removeKey: string) => {
    const params: any = {}
    searchParams.forEach((value: string | null, key: string) => {
        if (removeKey === key) return

        params[key] = value
    })

    return ({ ...params })
}
