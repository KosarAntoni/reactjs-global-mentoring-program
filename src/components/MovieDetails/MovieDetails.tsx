import React, { FC } from 'react'
import classNames from 'classnames'
import { formatGenres, numberToTime } from 'utilities'

import { MovieDetailsProps } from './MovieDetails.models'

const MovieDetails: FC<MovieDetailsProps> = ({
  title,
  overview,
  poster_path: posterPath,
  vote_average: voteAverage,
  genres, release_date:
  releaseDate,
  className,
  runtime
}) => {
  const year = new Date(releaseDate).getFullYear()

  return (
    <div className={classNames('movie-details', className)}>
      <img alt={title} src={posterPath} />

      <div>
        <div className='movie-details__heading'>
          <h2>
            {title}
          </h2>

          <span>
            {voteAverage}
          </span>
        </div>

        <p className='movie-details__genres'>
          {formatGenres(genres)}
        </p>

        <div className='movie-details__information'>
          <span>
            {year}
          </span>

          <span>
            {numberToTime(runtime)}
          </span>
        </div>

        <p className='movie-details__overview'>
          {overview}
        </p>
      </div>
    </div>
  )
}

export default MovieDetails
