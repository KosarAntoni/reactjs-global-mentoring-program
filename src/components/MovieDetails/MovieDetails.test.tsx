import React from 'react'
import { render, screen } from '@testing-library/react'
import { formatGenres } from 'utilities/formatGenres'
import { numberToTime } from 'utilities/formatTime'

import MovieDetails from './MovieDetails'
import { movieMock } from './MovieDetails.mock'

import '@testing-library/jest-dom'

test('the MovieDetails component renders correctly with required and optional props', () => {
  render(<MovieDetails className='test-class' {...movieMock} />)

  const node = screen.getByTestId('movie-details-node')

  // the component has been rendered
  expect(node).toBeInTheDocument()

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(node).toMatchSnapshot()

  // has image
  const imageNode = screen.getByAltText(movieMock.title)
  expect(imageNode).toBeInTheDocument()
  expect(imageNode).toHaveAttribute('src', movieMock.poster_path)

  // has correct rintime
  const runtimeNode = screen.getByText(numberToTime(movieMock.runtime))
  expect(runtimeNode).toBeInTheDocument()

  // has correct formated genres
  const genresNode = screen.getByText(formatGenres(movieMock.genres))
  expect(genresNode).toBeInTheDocument()

  // the comonent has correct classes
  expect(node).toHaveClass('movie-details test-class')
})
