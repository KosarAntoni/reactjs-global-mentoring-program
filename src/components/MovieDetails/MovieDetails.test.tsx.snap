// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the MovieDetails component renders correctly with required and optional props 1`] = `
<div
  class="movie-details test-class"
  data-testid="movie-details-node"
>
  <img
    alt="Fifty Shades Freed"
    src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
  />
  <div>
    <div
      class="movie-details__heading"
    >
      <h2>
        Fifty Shades Freed
      </h2>
      <span>
        6.1
      </span>
    </div>
    <p
      class="movie-details__genres"
    >
      Drama & Romance
    </p>
    <div
      class="movie-details__information"
    >
      <span>
        2018
      </span>
      <span>
        01h 46min
      </span>
    </div>
    <p
      class="movie-details__overview"
    >
      Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.
    </p>
  </div>
</div>
`;
